### DISCUSSION TITLE
Upsert readmodels

### DISCUSSION THREAD

**Discussion by peteglassey** [Eventuous/eventuous#287] [General]
Hi i cant seem to get an upsert to work for the mongo projector.

```
        On<LeaveDue>(c =>
            c.UpdateOne
                .Id(ctx => ctx.Message.LeavePeriodId.ToString())
                .Update((evt, update) =>
                    update
                        // rest removed for clarity
                        .Set(x => x.LeaveTypeDescription, evt.LeaveTypeDescription)));
```
There are 6 events in the stream.

I can see the checkpoint has been updated
![image](https://github.com/Eventuous/eventuous/assets/146303125/3b7d40a1-67ac-47a5-ab29-0b638e3b1bde)

but the readmodel never gets created.
![image](https://github.com/Eventuous/eventuous/assets/146303125/47a7e7cb-c988-41ba-ae9c-d760683034b0)

The https://eventuous.dev/docs/infra/mongodb mentions upserts are preferred and are on by default, but it doesnt seem to be working for me.

Any ideas?

**Comment by HeathLoganCampbell** [comment]
Wow man, same!

**Comment by peteglassey** [comment]
I realise i didnt provide a lot of context. 
https://github.com/petegee/Upsert/

Check this out - its a min app which i think shows what im seeing. The checkpoint document gets created and is correct, and i can see the logs saying its handling the events in the projector etc, but no readmodel gets created.

Just to be clear, i want the one HTTP POST to either create or update the read-model. I get the feeling im missing a peice of the puzzle?

**Reply by alexeyzimarev** [reply]
Good to have a repro, thanks for that

**Reply by alexeyzimarev** [reply]
Ok, I cloned the repo and called the API using Swagger two times with a different name, and two documents appeared in MongoDB. I changed the ESDB image tag to use latest arm64 image though because x64 images don't work on my machine.

From my side, everything works as expected.

**Reply by peteglassey** [reply]
thats interesting - it repros it for me, and some of my co-workers. What versions of mongo and esdb did you use?

**Reply by alexeyzimarev** [reply]
I use ESDB 23.6 ARM64 container and latest mongo container.

**Reply by peteglassey** [reply]
Looks like it was due to us using an older version of event store DB. When using ver 23.X and later, there seems to be no issue.

**Comment by alexeyzimarev** [comment]
The default options for update operations is using `IsUpsert` set to `true`. I need to check the tests, but I think they use `UpdateOne` and the documents are getting created, but now I am not so sure.

Here's the default options definition:
```csharp
        static UpdateOptions DefaultOptions => new() { IsUpsert = true };
```
so it's either not being picked up, or it's something else.

**Comment by discoaaron** [comment]
Hi Alexy

Here is a video of the issue, 
There are three events in event store when the application starts.
These are processed, checkpoints are created but no read model is created for any of them.
After this is complete I push another event into event store (via the post endpoint)
This event is processed and a read model is created.

Thoughts?

https://github.com/Eventuous/eventuous/assets/4475215/93de68de-c4b0-4ef0-9fb8-a1152ecc4eda

**Reply by alexeyzimarev** [reply]
It's really weird. You can of course try to set the options explicitly with `IsUpsert = true`.

**Reply by alexeyzimarev** [reply]
https://github.com/Eventuous/eventuous/assets/2821205/58ebb01d-f1d7-4766-8b7a-4a2a65e9383a

**Comment by peteglassey** [comment]
Hi Alexey. Thanks for the effort looking into this issue. I noticed that you are running ESDB 23.X where we (and in the our docker-compose in the repro repo) we're using 20.X. When i re-run using the 23.X container, it all works as it should and is perfectly inline with your video.

So, i guess the easy answer is to run ESDB 23.X - which is what we will do. Im hoping that this may provide some insight and help for you to try repro the issue with a 20.X container? 

https://github.com/petegee/Upsert/pull/1/files

**Reply by alexeyzimarev** [reply]
Maybe you can post as replies to my first response, as your initial question now has more questions as answers. I don't know why the issue happens with v20 as I never tested with that version. It's no longer supported, ESDB LTS support period is two years as far as I remember.

**Reply by alexeyzimarev** [reply]
In fact, as 23.10 is now out, 21.10 is also OOS
<img width="796" alt="image" src="https://github.com/Eventuous/eventuous/assets/2821205/8bdf817f-0ab8-4b4d-960a-211fa0c76578">

**Comment by peteglassey** [comment]
Looks like it was due to us using an older version of event store DB. When using ver 23.X and later, there seems to be no issue.
